Allow Promise to display error strings and not just error objects.

mikelambert opened the pull request.

facebook-github-bot-8 commented:
By analyzing the blame information on this pull request, we identified @davidaurelio and @danieldunderfelt to be potential reviewers.

ide commented:
Better to handle Errors specially and then have a fallback for all non-Error types ex: with https://github.com/thejameskyle/pretty-format

facebook-github-bot-3 commented:
@mikelambert updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9989/files/05ade5a0a89c7c3339b017150128fe4d2991de2a..215a41b28a130c3868365a346205bb06270fadfd)

mikelambert commented:
You want me to add another dependency to react-native? I can do that if desired, just figured a simple change that handled the (what I assumed, only) case of strings was a better and easier to get merged.

ide commented:
It's easier to get merged but more arbitrary and handling all values seems better to me.

davidaurelio commented on Libraries/Promise.js:
```diff
@@ -17,7 +17,10 @@ if (__DEV__) {
+      let {message = null, stack = null} = error;
+      if (typeof error === 'string' || error instanceof String) {
+        message = error;
+      }
```
What about `if (message == null) { message = String(error); }

davidaurelio commented on Libraries/Promise.js:
```diff
@@ -17,7 +17,10 @@ if (__DEV__) {
+      }
       const warning =
         `Possible Unhandled Promise Rejection (id: ${id}):\n` +
         (message == null ? '' : `${message}\n`) +
```
or here: `${message == null ? String(error) : message}\n`

davidaurelio commented:
`pretty-format` would work, too, of course

mikelambert commented:
Thanks David, I'm partial to String(error) until we find errors raised that warrant more, but let me know if you prefer pretty-format, and I'll make the change.

And sorry if I'm mistaken @ide , but I'm assuming you're an external dev chiming in but not a core RN committer, so am deferring to @davidaurelio on the pretty-format dependency. Otherwise I'm not sure how to weight your preference on this change. :)

davidaurelio commented:
@mikelambert sounds good. @ide is as core as it gets 😉

facebook-github-bot-1 commented:
@mikelambert updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9989/files/215a41b28a130c3868365a346205bb06270fadfd..ff7c85edf164b331abfbb931878b9666b9ad1e50)

facebook-github-bot-8 commented:
@mikelambert updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9989/files/ff7c85edf164b331abfbb931878b9666b9ad1e50..215a41b28a130c3868365a346205bb06270fadfd)

mikelambert commented:
Sorry for the fb-bot spam, I'm dealing with this git learning curve on my end, and screwed up my branches. This should be back to the original pull again (sorry, haven't had made time yet to make the pretty-format change and test it yet)

facebook-github-bot-8 commented:
@mikelambert updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9989/files/215a41b28a130c3868365a346205bb06270fadfd..6f6403d7e334e6376dea6ae0943e4ad80d9d66d7)

mikelambert commented:
Sorry for the delay @davidaurelio and @ide, I have updated this pull request to use pretty-format and tested that it works.

ide commented on Libraries/Promise.js:
```diff
@@ -11,13 +11,17 @@
     allRejections: true,
     onUnhandled: (id, error = {}) => {
-      const {message = null, stack = null} = error;
+      let {message = null, stack = null} = error;
```
It'd be useful to handle `null` errors and include the error type (TypeError, ReferenceError, etc) too while we're at it. Can you change this method to this implementation?

```
onUnhandled: (id, error: any) => {
  let message: string;
  let stack: ?string;
  
  const stringValue = Object.prototype.toString.call(error);
  if (stringValue === '[object Error]') {
    message = Error.prototype.toString.call(error);
    stack = error.stack;
  } else {
    message = prettyFormat(error);
  }
  
  const warning =
    `Possible Unhandled Promise Rejection (id: ${id}):\n` +
    `${message}\n` +
    (stack == null ? '' : stack);
  console.warn(warning);
}
```

hramos commented:
@mikelambert still planning to ship this?

mikelambert commented:
Sorry @hramos. I've just updated with the latest set of changes @ide suggested. I've tested that it works and logs correctly with:
throw 'str';
throw new Error('str');
throw new RangeError('str');

ide commented:
@mikelambert thanks, this is robust and simple! The CircleCI tests look like they are failing because of a JDK issue, unrelated to this PR. I don't think there are automated tests for promise rejections so  we aren't missing much signal from CircleCI plus you said you manually tested it.

@hramos Could you help import this PR? It adds a dependency (https://github.com/thejameskyle/pretty-format).

mikelambert commented:
What is the next step for this change? @hramos imported it 18 days ago, but it is not yet in master. Is it sitting internally waiting in phabricator for review?

mikelambert commented:
Ping?

mikelambert commented:
Ping @ide @hramos ?

ide commented:
@ericvicenti can you import and commit this?

mikelambert commented:
I've updated and fixed a merge conflict on `package.json`, so hopefully this should be safe to try again, Mr. @facebook-github-bot .

Sorry about the commit-log mess. My git-foo is not good enough to rebase it properly without messing up the entire branch. Will your phabricator import collapse it all? If not...I can make an attempt at rebasing and collapsing it, and worst-case submit a new clean pull request.

Thanks!

mikelambert commented:
Ping @ericvicenti since I re-merged master, and suspect it needs re-triggering on your end.

mikelambert commented:
@hramos and @ericvicenti to clarify if this is something I need to act on or not...
- It passes all checks according to github
- I just recently merged to github master, dunno if I already need to do it again?
- I'm assuming it instead must be breaking some internal test?

hramos commented:
@mikelambert nothing needed on your part. Anything that touches package.json requires a manual import on our side. We have already done that part, but the internal diff is failing some internal tests. It looks like the failures are transient and unrelated. I'll kick it off again and see if it passes tests again.

hramos commented:
Finally!
